DRAFT: "AutoRouter" to reduce router initialization boilerplate

kwhitley opened the pull request:
### Description
Adds `flow(router: Router, options?: FlowOptions): Response` to reduce boilerplate.  This function makes the following assumptions (with override controls):
- Default response formatter as `json`
- Includes default 404 handler on missed route
- Includes default 500 error handler on uncaught errors

```ts
// BEFORE/WITHOUT FLOW
import { Router, json, error } from 'itty-router'

const router = Router()

router
  .get('/my-items', () => [1, 2, 3])
  .get('*', () => error(404))

export default {
  fetch: (...args) => router
                        .handle(...args)
                        .then(json)
                        .catch(error)
}
```

```ts
// AFTER/WITH FLOW
import { Router, flow } from 'itty-router'

const router = Router()

router.get('/my-items', () => [1, 2, 3])

export default {
  fetch: flow(router)
}

// or even

export default flow(router)
```

```ts
// WITH FLOW, OVERRIDES (including CORS)
import { Router, flow, error } from 'itty-router'

const router = Router()

router.get('/my-items', () => [1, 2, 3])

export default {
  fetch: flow(router, {
    cors: true,
    notFound: (req) => error(404, `The path "${req.url}" was not found.`),
  })
}
```

#### API (options)
| Option Name | Default | Valid Choices |
| --- | --- | --- |
| cors | `false` | `false` (off/ignore), `true` (default `createCors()`), `CorsOptions` |
| format | `json` | `false` (off/ignore), `Function<Response>` |
| errors | `error` | `false` (off/ignore), `Function<Response>` |
| notFound | `() => error(404)` | `false` (off/ignore), `Function<Response>` |

### Type of Change (select one and follow subtasks)
- [ ] Documentation (README.md)
- [ ] Maintenance or repo-level work (e.g. linting, build, tests, refactoring, etc.)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
  - [x] I have included test coverage
- [ ] Breaking change (fix or feature that would cause existing functionality/userland code to not work as expected)
  - [ ] Explain why a breaking change is necessary:
- [x] This change requires (or is) a documentation update
  - [ ] I have added necessary local documentation (if appropriate)
  - [ ] I have added necessary [itty.dev](https://github.com/kwhitley/itty.dev) documentation (if appropriate)

kwhitley committed:
testing flow

kwhitley committed:
released v4.1.0-next.0 - initial flow release

kwhitley committed:
released v4.1.0-next.1 - flow should be exported from index

kwhitley committed:
added flow examples

kwhitley committed:
released v4.1.0-next.2 - updated package.json

kwhitley committed:
released v4.1.0-next.3 - changing API of flow

kwhitley committed:
released v4.1.0-next.4 - code golf optimization in flow

kwhitley committed:
flow work

kwhitley committed:
flow working with tests - TODO: finalize API

kwhitley committed:
fixed merge conflicts

kwhitley committed:
ts-error should have description

kwhitley committed:
fixing lint errors

kwhitley committed:
Merge remote-tracking branch 'origin/v4.x' into flow

kwhitley committed:
update to CHANGELOG

kwhitley committed:
Merge remote-tracking branch 'origin/v4.x' into flow

kwhitley committed:
deps update

kwhitley committed:
update actions to use Node v20

kwhitley committed:
mock Response for websocket test

kwhitley committed:
fix node version

kwhitley committed:
flow signatures added and tests fixed

kwhitley committed:
remove support for router.fetch

kwhitley committed:
resolved merge conflicts

kwhitley committed:
deps update

kwhitley committed:
testing after and before handlers in flow

kwhitley committed:
tweak bun flow advanced example

kwhitley committed:
default CORS option is now to allow all methods

kwhitley committed:
CHANGELOG added for createCors fix

kwhitley committed:
fixed #202 - typescript generics not being used at router level if IRequest used, but Args changed

kwhitley committed:
fixed request/args TS flow?!

kwhitley committed:
modified changelog and Router types

kwhitley committed:
checking in

kwhitley committed:
updating flow spec

kwhitley committed:
after honors return if given

kwhitley committed:
shorter way to write the after bit

Crisfole commented on src/Router.ts:
```diff
@@ -27,70 +25,63 @@ export type RouterOptions = {
-export type RouteHandler<I = IRequest, A extends any[] = any[]> = {
-  (request: I, ...args: A): any
+export type RouteHandler<R = IRequest, Args = any[]> = {
+  // @ts-expect-error - TS never likes this syntax
```
Doesn’t like the syntax because you need [] array type for spread args.

Crisfole replied:
You’d need an extends Array<any> in your type handler here, I think

Crisfole committed:
Bugfix/ts expect error is a bug (#205)

* Add the required `extends` calls to fix the ts-expect-error

* Get rid of the other typescript ignores

* Replace one  last expect-error with an `as`

In this case the 'as' is asserting "I know better than the compiler" and it
is both true and more accurate than saying "I expect typescript to error out
here." As an added benefit, `as` does more type checking than 'expect error',
so you get just a small bit more type safety.

kwhitley committed:
merge conflicts

kwhitley committed:
AutoRouter

kwhitley committed:
passing all tests

Crisfole commented on src/createCors.ts:
```diff
@@ -7,13 +7,13 @@ export type CorsOptions = {
-export interface CorsFns {
-  corsify(req: Request): Response;
-  preflight(r: IRequest): Response;
+export interface CorsHandlers {
```
MUCH better name. Thanks for the fix.

Crisfole commented on src/createCors.ts:
```diff
@@ -7,13 +7,13 @@ export type CorsOptions = {
-  preflight(r: IRequest): Response;
+export interface CorsHandlers {
+  corsify(response: Response): any
+  preflight(Request: IRequest): any
```
Never use a direct `IRequest` anywhere but as a default for a generic type. It'd be best if you setup a linter for this, tbh. It's the source of 99% of your type bugs.

Crisfole commented on src/createCors.ts:
```diff
@@ -38,20 +38,20 @@ export const createCors = (options: CorsOptions = {}): CorsFns => {
 
   // Pre-flight function.
-  const preflight = (r: IRequest) => {
+  const preflight = (request: IRequest): any => {
```
`createCors` needs to be generic to get the types to behave.

Crisfole commented on src/flow.ts:
```diff
@@ -20,7 +20,7 @@ export type FlowOptions = {
 }
 
-export type Flowed = (request: RequestLike, ...extra: any[]) => Promise<any>
+export type Flowed = (request: IRequest, ...extra: any[]) => Promise<any>
```
This has to be generic to get the handler types working nicely.

Crisfole commented on src/flow.ts:
```diff
@@ -20,7 +20,7 @@ export type FlowOptions = {
 
-export type Flowed = (request: RequestLike, ...extra: any[]) => Promise<any>
+export type Flowed = (request: IRequest, ...extra: any[]) => Promise<any>
 export type FlowedAndFetch = Flowed & { fetch: Flowed }
```
This is a real weird type. It's a fetch function that owns a ... fetch function?

Crisfole reviewed (changes requested):
@kwhitley this is mostly an improvement, but you're missing more generics. That's what causes some 95% of your ts-ignore-error and `any` usages.

Pretty much every `any` you have can be replaced 1:1 with a generic type parameter for significantly better type signatures.

kwhitley commented:
Adding to the discussion here... `flow` has been shelved for an arguably better router approach.  Notes to follow.

kwhitley committed:
conflicts

kwhitley commented:
Killed in favor of #222
